File: Android App/app/src/main/java/com/ikuteam/notestn/data/joplin/JoplinCloudApi.kt
package com.ikuteam.notestn.data.joplin

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * Client for Joplin Cloud's sync API (the "JoplinServer" protocol — Joplin Cloud is
 * just a hosted instance of Joplin Server). Phase 1 only needs login; item sync
 * (api/items, api/batch_items, delta cursor) is a later phase.
 *
 * Protocol reference: packages/lib/JoplinServerApi.ts in this repo's Joplin monorepo.
 */
object JoplinCloudApi {

    private const val BASE_URL = "https://api.joplincloud.com"
    private val JSON_MEDIA_TYPE = "application/json".toMediaType()
    private val json = Json { ignoreUnknownKeys = true }

    // Real Joplin clients send this on every request (see JoplinServerApi.ts's
    // exec_() — "Need server 2.6 for new lock support"). We don't do sync locks,
    // but the server may use this header to decide how much of its newer
    // processing (e.g. change-log registration for delta) to apply, so send it
    // too rather than looking like a legacy/unversioned client.
    private val client = OkHttpClient.Builder()
        .addInterceptor { chain ->
            chain.proceed(chain.request().newBuilder().header("X-API-MIN-VERSION", "2.6.0").build())
        }
        .build()

    @Serializable
    data class LoginResult(val id: String, val userId: String)

    @Serializable
    private data class LoginRequest(val email: String, val password: String)

    @Serializable
    private data class SessionResponse(val id: String, val user_id: String)

    @Serializable
    private data class ErrorResponse(val error: String? = null, val message: String? = null)

    sealed class LoginError : Exception() {
        object InvalidCredentials : LoginError()
        data class Other(override val message: String) : LoginError()
        data class Network(override val cause: Throwable) : LoginError()
    }

    /** POST api/sessions — email/password login, returns a session id used as the
     * X-API-AUTH header on every later sync request. */
    suspend fun login(email: String, password: String): Result<LoginResult> = withContext(Dispatchers.IO) {
        val body = json.encodeToString(LoginRequest.serializer(), LoginRequest(email, password))

        val request = Request.Builder()
            .url("$BASE_URL/api/sessions")
            .post(body.toRequestBody(JSON_MEDIA_TYPE))
            .build()

        try {
            client.newCall(request).execute().use { response ->
                val responseBody = response.body?.string().orEmpty()
                if (response.isSuccessful) {
                    val session = json.decodeFromString<SessionResponse>(responseBody)
                    Result.success(LoginResult(id = session.id, userId = session.user_id))
                } else if (response.code == 403 || response.code == 401) {
                    Result.failure(LoginError.InvalidCredentials)
                } else {
                    val error = runCatching { json.decodeFromString<ErrorResponse>(responseBody) }.getOrNull()
                    Result.failure(LoginError.Other(error?.message ?: error?.error ?: "Login failed (HTTP ${response.code})"))
                }
            }
        } catch (t: Throwable) {
            Result.failure(LoginError.Network(t))
        }
    }

    // MARK: - Sync (Phase 2, pull-only)

    @Serializable
    data class DeltaChange(
        val id: String,
        // The item's filename, e.g. "<32-char-id>.md" — this, not `id`/`item_id` (the
        // server's internal change-log row id), is what content fetches are keyed on.
        val item_name: String? = null,
        // Numeric ChangeType from Joplin's source: 1 = create, 2 = update, 3 = delete.
        val type: Int = 1,
    )

    @Serializable
    data class DeltaResponse(
        val items: List<DeltaChange> = emptyList(),
        val has_more: Boolean = false,
        val cursor: String? = null,
    )

    sealed class SyncApiError : Exception() {
        object Unauthorized : SyncApiError()
        data class Other(override val message: String) : SyncApiError()
        data class Network(override val cause: Throwable) : SyncApiError()
    }

    /** GET api/items/root:/:/delta — root:/:/  is the literal sentinel for "account
     * root" per file-api-driver-joplinServer.ts. Paginate with the returned cursor
     * until has_more is false. */
    suspend fun delta(sessionId: String, cursor: String?): Result<DeltaResponse> = withContext(Dispatchers.IO) {
        val url = "$BASE_URL/api/items/root:/:/delta" + if (cursor != null) "?cursor=$cursor" else ""
        authorizedGet(url, sessionId) { response -> json.decodeFromString<DeltaResponse>(response.body?.string().orEmpty()) }
    }

    /** GET api/items/root:/{itemName}:/content — raw serialized item text (title,
     * body, metadata footer — see JoplinItemParser). */
    suspend fun itemContent(sessionId: String, itemName: String): Result<String> = withContext(Dispatchers.IO) {
        val url = "$BASE_URL/api/items/root:/$itemName:/content"
        authorizedGet(url, sessionId) { response -> response.body?.string().orEmpty() }
    }

    /** GET api/items/root:/.resource/{resourceId}:/content — the resource's raw binary
     * blob, stored as a separate item from its `{resourceId}.md` metadata (see
     * Dirnames.Resources / resourceRemotePath() in packages/lib). Must be read as
     * bytes, not text, or the image data gets corrupted by string decoding. */
    suspend fun resourceBlob(sessionId: String, resourceId: String): Result<ByteArray> = withContext(Dispatchers.IO) {
        val url = "$BASE_URL/api/items/root:/.resource/$resourceId:/content"
        authorizedGet(url, sessionId) { response -> response.body?.bytes() ?: ByteArray(0) }
    }

    private val OCTET_STREAM = "application/octet-stream".toMediaType()

    /** PUT api/items/root:/{itemName}:/content — creates OR updates an item; Joplin
     * Server has no separate create call, a PUT at this path upserts whatever's there. */
    suspend fun putItemContent(sessionId: String, itemName: String, content: ByteArray): Result<Unit> =
        withContext(Dispatchers.IO) {
            val url = "$BASE_URL/api/items/root:/$itemName:/content"
            val request = Request.Builder()
                .url(url)
                .header("X-API-AUTH", sessionId)
                .put(content.toRequestBody(OCTET_STREAM))
                .build()
            executeForResult(request, url) { }
        }

    /** PUT api/items/root:/.resource/{resourceId}:/content — uploads a resource's raw
     * binary blob, the counterpart to resourceBlob()'s GET. Separate path from the
     * resource's own `{resourceId}.md` metadata item (pushed via putItemContent). */
    suspend fun putResourceBlob(sessionId: String, resourceId: String, content: ByteArray): Result<Unit> =
        withContext(Dispatchers.IO) {
            val url = "$BASE_URL/api/items/root:/.resource/$resourceId:/content"
            val request = Request.Builder()
                .url(url)
                .header("X-API-AUTH", sessionId)
                .put(content.toRequestBody(OCTET_STREAM))
                .build()
            executeForResult(request, url) { }
        }

    /** DELETE api/items/root:/{itemName}: — no `/content` suffix, unlike get/put. */
    suspend fun deleteItem(sessionId: String, itemName: String): Result<Unit> = withContext(Dispatchers.IO) {
        val url = "$BASE_URL/api/items/root:/$itemName:"
        val request = Request.Builder()
            .url(url)
            .header("X-API-AUTH", sessionId)
            .delete()
            .build()
        executeForResult(request, url) { }
    }

    private fun <T> executeForResult(request: Request, url: String, parse: (okhttp3.Response) -> T): Result<T> {
        return try {
            client.newCall(request).execute().use { response ->
                when {
                    // A delete for an item that's already gone (e.g. retried after a
                    // previous run succeeded but the local "pending delete" wasn't
                    // cleared) is not an error — treat 404 as success too.
                    response.isSuccessful || response.code == 404 -> Result.success(parse(response))
                    response.code == 401 || response.code == 403 -> Result.failure(SyncApiError.Unauthorized)
                    else -> Result.failure(SyncApiError.Other("Request failed (HTTP ${response.code}): $url"))
                }
            }
        } catch (t: Throwable) {
            Result.failure(SyncApiError.Network(t))
        }
    }

    private fun <T> authorizedGet(url: String, sessionId: String, parse: (okhttp3.Response) -> T): Result<T> {
        val request = Request.Builder()
            .url(url)
            .header("X-API-AUTH", sessionId)
            .get()
            .build()

        return try {
            client.newCall(request).execute().use { response ->
                when {
                    response.isSuccessful -> Result.success(parse(response))
                    response.code == 401 || response.code == 403 -> Result.failure(SyncApiError.Unauthorized)
                    else -> Result.failure(SyncApiError.Other("Request failed (HTTP ${response.code}): $url"))
                }
            }
        } catch (t: Throwable) {
            Result.failure(SyncApiError.Network(t))
        }
    }
}

File: Mac/NotesTN/NotesTN/Sync/JoplinCloudApi.swift
import Foundation

/// Client for Joplin Cloud's sync API (the "JoplinServer" protocol — Joplin Cloud is
/// just a hosted instance of Joplin Server). Phase 1 only needs login; item sync
/// (api/items, api/batch_items, delta cursor) is a later phase.
///
/// Protocol reference: packages/lib/JoplinServerApi.ts in this repo's Joplin monorepo.
enum JoplinCloudApi {

    private static let baseURL = URL(string: "https://api.joplincloud.com")!

    // Real Joplin clients send this on every request (see JoplinServerApi.ts's
    // exec_() — "Need server 2.6 for new lock support"). We don't do sync locks,
    // but the server may use this header to decide how much of its newer
    // processing (e.g. change-log registration for delta) to apply, so send it
    // too rather than looking like a legacy/unversioned client.
    private static let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.httpAdditionalHeaders = ["X-API-MIN-VERSION": "2.6.0"]
        return URLSession(configuration: config)
    }()

    struct LoginResult {
        let sessionId: String
        let userId: String
    }

    enum LoginError: Error, LocalizedError {
        case invalidCredentials
        case other(String)
        case network(Error)

        var errorDescription: String? {
            switch self {
            case .invalidCredentials: return "Incorrect email or password."
            case .other(let message): return message
            case .network: return "Couldn't reach Joplin Cloud. Check your connection."
            }
        }
    }

    private struct LoginRequestBody: Encodable {
        let email: String
        let password: String
    }

    private struct SessionResponse: Decodable {
        let id: String
        let userId: String

        enum CodingKeys: String, CodingKey {
            case id
            case userId = "user_id"
        }
    }

    private struct ErrorResponse: Decodable {
        let error: String?
        let message: String?
    }

    /// POST api/sessions — email/password login, returns a session id used as the
    /// X-API-AUTH header on every later sync request.
    static func login(email: String, password: String) async -> Result<LoginResult, LoginError> {
        var request = URLRequest(url: baseURL.appendingPathComponent("api/sessions"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONEncoder().encode(LoginRequestBody(email: email, password: password))

        do {
            let (data, response) = try await session.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse else {
                return .failure(.other("Unexpected response from server."))
            }

            if httpResponse.statusCode == 200 || httpResponse.statusCode == 201 {
                let session = try JSONDecoder().decode(SessionResponse.self, from: data)
                return .success(LoginResult(sessionId: session.id, userId: session.userId))
            } else if httpResponse.statusCode == 401 || httpResponse.statusCode == 403 {
                return .failure(.invalidCredentials)
            } else {
                let errorBody = try? JSONDecoder().decode(ErrorResponse.self, from: data)
                let message = errorBody?.message ?? errorBody?.error ?? "Login failed (HTTP \(httpResponse.statusCode))"
                return .failure(.other(message))
            }
        } catch {
            return .failure(.network(error))
        }
    }

    // MARK: - Sync (Phase 2, pull-only)

    struct DeltaChange: Decodable {
        let id: String
        // The item's filename, e.g. "<32-char-id>.md" — this, not `id`/`item_id` (the
        // server's internal change-log row id), is what content fetches are keyed on.
        let itemName: String?
        // Numeric ChangeType from Joplin's source: 1 = create, 2 = update, 3 = delete.
        let type: Int

        enum CodingKeys: String, CodingKey {
            case id
            case itemName = "item_name"
            case type
        }

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            id = try container.decode(String.self, forKey: .id)
            itemName = try container.decodeIfPresent(String.self, forKey: .itemName)
            type = try container.decodeIfPresent(Int.self, forKey: .type) ?? 1
        }
    }

    struct DeltaResponse: Decodable {
        let items: [DeltaChange]
        let hasMore: Bool
        let cursor: String?

        enum CodingKeys: String, CodingKey {
            case items
            case hasMore = "has_more"
            case cursor
        }
    }

    enum SyncApiError: Error, LocalizedError {
        case unauthorized
        case other(String)
        case network(Error)

        var errorDescription: String? {
            switch self {
            case .unauthorized: return "Joplin Cloud session expired — please log in again."
            case .other(let message): return message
            case .network: return "Couldn't reach Joplin Cloud. Check your connection."
            }
        }
    }

    /// GET api/items/root:/:/delta — root:/:/  is the literal sentinel for "account
    /// root" per file-api-driver-joplinServer.ts. Paginate with the returned cursor
    /// until hasMore is false.
    static func delta(sessionId: String, cursor: String?) async -> Result<DeltaResponse, SyncApiError> {
        var urlString = baseURL.appendingPathComponent("api/items/root:/:/delta").absoluteString
        if let cursor { urlString += "?cursor=\(cursor)" }
        return await authorizedGet(urlString: urlString, sessionId: sessionId) { data in
            try JSONDecoder().decode(DeltaResponse.self, from: data)
        }
    }

    /// GET api/items/root:/{itemName}:/content — raw serialized item text (title,
    /// body, metadata footer — see JoplinItemParser).
    static func itemContent(sessionId: String, itemName: String) async -> Result<String, SyncApiError> {
        let urlString = baseURL.appendingPathComponent("api/items/root:/\(itemName):/content").absoluteString
        return await authorizedGet(urlString: urlString, sessionId: sessionId) { data in
            String(data: data, encoding: .utf8) ?? ""
        }
    }

    /// GET api/items/root:/.resource/{resourceId}:/content — the resource's raw binary
    /// blob, stored as a separate item from its `{resourceId}.md` metadata (see
    /// Dirnames.Resources / resourceRemotePath() in packages/lib).
    static func resourceBlob(sessionId: String, resourceId: String) async -> Result<Data, SyncApiError> {
        let urlString = baseURL.appendingPathComponent("api/items/root:/.resource/\(resourceId):/content").absoluteString
        return await authorizedGet(urlString: urlString, sessionId: sessionId) { data in data }
    }

    /// PUT api/items/root:/{itemName}:/content — creates OR updates an item; Joplin
    /// Server has no separate create call, a PUT at this path upserts whatever's there.
    static func putItemContent(sessionId: String, itemName: String, content: Data) async -> Result<Void, SyncApiError> {
        let urlString = baseURL.appendingPathComponent("api/items/root:/\(itemName):/content").absoluteString
        guard let url = URL(string: urlString) else { return .failure(.other("Invalid URL: \(urlString)")) }
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue(sessionId, forHTTPHeaderField: "X-API-AUTH")
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        request.httpBody = content
        return await executeForResult(request, urlString: urlString)
    }

    /// PUT api/items/root:/.resource/{resourceId}:/content — uploads a resource's raw
    /// binary blob, the counterpart to resourceBlob()'s GET. Separate path from the
    /// resource's own `{resourceId}.md` metadata item (pushed via putItemContent).
    static func putResourceBlob(sessionId: String, resourceId: String, content: Data) async -> Result<Void, SyncApiError> {
        let urlString = baseURL.appendingPathComponent("api/items/root:/.resource/\(resourceId):/content").absoluteString
        guard let url = URL(string: urlString) else { return .failure(.other("Invalid URL: \(urlString)")) }
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue(sessionId, forHTTPHeaderField: "X-API-AUTH")
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        request.httpBody = content
        return await executeForResult(request, urlString: urlString)
    }

    /// DELETE api/items/root:/{itemName}: — no `/content` suffix, unlike get/put.
    static func deleteItem(sessionId: String, itemName: String) async -> Result<Void, SyncApiError> {
        let urlString = baseURL.appendingPathComponent("api/items/root:/\(itemName):").absoluteString
        guard let url = URL(string: urlString) else { return .failure(.other("Invalid URL: \(urlString)")) }
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.setValue(sessionId, forHTTPHeaderField: "X-API-AUTH")
        return await executeForResult(request, urlString: urlString)
    }

    private static func executeForResult(_ request: URLRequest, urlString: String) async -> Result<Void, SyncApiError> {
        do {
            let (_, response) = try await session.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse else {
                return .failure(.other("Unexpected response from server."))
            }
            if httpResponse.statusCode == 401 || httpResponse.statusCode == 403 {
                return .failure(.unauthorized)
            }
            // A delete for an item that's already gone (e.g. retried after a previous
            // run succeeded but the local "pending delete" wasn't cleared) is not an
            // error — treat 404 as success too.
            guard (200...299).contains(httpResponse.statusCode) || httpResponse.statusCode == 404 else {
                return .failure(.other("Request failed (HTTP \(httpResponse.statusCode)): \(urlString)"))
            }
            return .success(())
        } catch {
            return .failure(.network(error))
        }
    }

    private static func authorizedGet<T>(
        urlString: String,
        sessionId: String,
        parse: (Data) throws -> T
    ) async -> Result<T, SyncApiError> {
        guard let url = URL(string: urlString) else { return .failure(.other("Invalid URL: \(urlString)")) }
        var request = URLRequest(url: url)
        request.setValue(sessionId, forHTTPHeaderField: "X-API-AUTH")

        do {
            let (data, response) = try await session.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse else {
                return .failure(.other("Unexpected response from server."))
            }
            if httpResponse.statusCode == 401 || httpResponse.statusCode == 403 {
                return .failure(.unauthorized)
            }
            guard (200...299).contains(httpResponse.statusCode) else {
                return .failure(.other("Request failed (HTTP \(httpResponse.statusCode)): \(urlString)"))
            }
            do {
                return .success(try parse(data))
            } catch {
                return .failure(.other("Failed to parse response: \(error.localizedDescription)"))
            }
        } catch {
            return .failure(.network(error))
        }
    }
}
